Finish API/State Changes

Myoldmopar opened the pull request:
Pull request overview
---------------------
This pull request future-proofs the API calls in relation to the new `state` object.  This will break API calls, which is why I am doing it now, in the self-specified grace period for major API changes during this first release cycle after the API was added.

This also cleans up and really improves the API documentation.  I intend on making an "application guide" of some form to help guide users to using the API and Python Plugin system as well, but that can be closer to release time.

Myoldmopar committed:
State now passed essentially everywhere through API and Plugin calls

Myoldmopar committed:
Move state functionality to separate part of API

Myoldmopar committed:
Docs tweaks, more to come

Myoldmopar committed:
Dont recreate Python strings each call

Myoldmopar committed:
Flesh out documentation on API and plugin code

Myoldmopar commented:
Wonderfully clean on CI.  I'll give a walkthrough of the changes I am making here.

Myoldmopar commented on cmake/PythonSetupAPIinBuild.cmake:
```diff
@@ -21,4 +21,5 @@ configure_file( "${API_SOURCE_DIR}/api.py" "${API_TARGET_DIR}/api.py" )
 configure_file( "${API_SOURCE_DIR}/func.py" "${API_TARGET_DIR}/func.py" )
 configure_file( "${API_SOURCE_DIR}/runtime.py" "${API_TARGET_DIR}/runtime.py" )
 configure_file( "${API_SOURCE_DIR}/plugin.py" "${API_TARGET_DIR}/plugin.py" )
+configure_file( "${API_SOURCE_DIR}/state.py" "${API_TARGET_DIR}/state.py" )
```
The new `state` object really needs to stand alone at the core of the EnergyPlus API, so I moved it to its own file.

Myoldmopar commented on src/EnergyPlus/PluginManager.cc:
```diff
@@ -105,13 +105,18 @@ namespace PluginManagement {
     void (*EP_Py_SetPythonHome)(const wchar_t *) = nullptr;
+    PyObjectWrap (*EP_Py_BuildValue)(const char *) = nullptr;
+    PyObjectWrap (*EP_PyLong_FromVoidPtr)(void *) = nullptr;
+    PyObjectWrap (*EP_PyString_FromString)(const char *) = nullptr;
```
We are now passing a pointer to the current `state` instance into the plugin override functions, so that those functions can then pass that state around as well.  Needed to add a few new functions to call into the Python DLL appropriately.

Myoldmopar commented on src/EnergyPlus/PluginManager.cc:
```diff
@@ -312,6 +327,21 @@ namespace PluginManagement {
+        if (!EP_PyString_FromString) {
+            ShowSevereError("Could not load Python function: EP_Wrap_PyString_FromString");
+            dll_errors = true;
+        }
```
Windows CI seemed to have no problem with these added calls.

Myoldmopar commented on src/EnergyPlus/PluginManager.cc:
```diff
@@ -416,6 +452,15 @@ namespace PluginManagement {
+        checkWrapperDLLFunction(err, "EP_Wrap_PyLong_FromVoidPtr", dll_errors);
+        *(void **)(&EP_PyString_FromString) = dlsym(wrapperDLLHandle, "EP_Wrap_PyString_FromString");
+        err = dlerror();
+        checkWrapperDLLFunction(err, "EP_Wrap_PyString_FromString", dll_errors);
```
Linux/Mac also liked this.

Myoldmopar commented on src/EnergyPlus/PluginManager.cc:
```diff
@@ -454,7 +499,7 @@ namespace PluginManagement {
         for (auto &plugin : plugins) {
             if (plugin.runDuringWarmup || !DataGlobals::WarmupFlag) {
-                bool const didOneRun = plugin.run(iCalledFrom);
+                bool const didOneRun = plugin.run(state, iCalledFrom);
```
Since we pass the current `state` to the plugin functions, we have to pass `state` around to a few new places in the code -- not many though.

Myoldmopar commented on src/EnergyPlus/PluginManager.cc:
```diff
@@ -1226,38 +1271,55 @@ namespace PluginManagement {
                 std::string functionName = (*EP_PyUnicode_AsUTF8)(item);
                 if (functionName == this->sHookBeginNewEnvironment) {
                     this->bHasBeginNewEnvironment = true;
+                    this->pBeginNewEnvironment = (*EP_PyString_FromString)(functionName.c_str());
```
Store `PyObject` versions of the function names just once to avoid string operations later.

Myoldmopar commented on src/EnergyPlus/PluginManager.cc:
```diff
@@ -1275,91 +1337,129 @@ namespace PluginManagement {
 #if LINK_WITH_PYTHON == 1
         (*EP_Py_DECREF)(this->pClassInstance);
         (*EP_Py_DECREF)(this->pModule); // PyImport_Import returns a new reference, decrement it
+        if (this->bHasBeginNewEnvironment) (*EP_Py_DECREF)(this->pBeginNewEnvironment);
```
Decrement the reference count for "new" PyObject instances if they exist.

Myoldmopar commented on src/EnergyPlus/PluginManager.cc:
```diff
@@ -1275,91 +1337,129 @@ namespace PluginManagement {
+        //static const PyObjectWrap oneArgObjFormat = (*EP_Py_BuildValue)("O");
+        PyObjectWrap pStateInstance = (*EP_PyLong_FromVoidPtr)((void*)&state);
+        PyObjectWrap pFunctionResponse = (*EP_PyObject_CallMethod2ObjArg)(this->pClassInstance, pFunctionName, pStateInstance, nullptr);
+        (*EP_Py_DECREF)(pStateInstance);
```
Call the appropriate Python plugin function and pass state to it.

Myoldmopar commented on src/EnergyPlus/PluginManager.hh:
```diff
@@ -99,43 +95,65 @@ namespace PluginManagement {
+        bool bHasEndOfZoneSizing = false;
+        bool bHasEndOfSystemSizing = false;
+        bool bHasAfterComponentInputReadIn = false;
+        bool bHasUserDefinedComponentModel = false;
```
I think I could actually get rid of these booleans now, and just use the PyObject variables to check for not `nullptr`, or _truthy._

Myoldmopar commented on src/EnergyPlus/PythonLibWrapper.hh:
```diff
@@ -99,6 +101,9 @@ extern "C" {
     PYTHONWRAP_API void EP_Wrap_Py_SetPythonHome(const wchar_t *home);
+    PYTHONWRAP_API PyObject * EP_Wrap_Py_BuildValue(const char *format);
+    PYTHONWRAP_API PyObject * EP_Wrap_PyLong_FromVoidPtr(void *v);
+    PYTHONWRAP_API PyObject * EP_Wrap_PyString_FromString(const char *s);
```
New functions added to the Python shared library intermediary.

Myoldmopar commented on src/EnergyPlus/api/api.py:
```diff
@@ -67,14 +68,10 @@ def __init__(self, running_as_python_plugin: bool = False):
-                api_version_defined_here, api_version_from_ep
-            ))
+        # self.state provides access to the main EnergyPlus state management class, instantiated and ready to go
+        self.state = State(self.api)
```
Deferred calling the API version check because it needs `state`, and we can't get a `state` until after the API is created...

Myoldmopar commented on src/EnergyPlus/api/datatransfer.cc:
```diff
@@ -60,7 +61,7 @@
 #include <EnergyPlus/PluginManager.hh>
 
-char * listAllAPIDataCSV() {
+char * listAllAPIDataCSV(EnergyPlusState) {
```
Added `state` as an argument to all the API functions to future proof them.  Right now most of the functions don't end up using it, but they will once we stop getting everything off the global program state.

Myoldmopar commented on src/EnergyPlus/api/plugin.py:
```diff
@@ -49,53 +47,53 @@ def _detect_overridden(self) -> List[str]:
         return diff
 
-    def on_begin_new_environment(self) -> int:
+    def on_begin_new_environment(self, state) -> int:
```
Added state as an argument to the plugin functions...all user-overridden functions will need to accept this argument, and then they can use it to call back into the API.

Myoldmopar commented on src/EnergyPlus/api/plugin_tester.py:
```diff
@@ -200,7 +200,8 @@ def plugin_file_tester(file_path):
                             try:
-                                response = method_to_call()
+                                dummy_state_arg = 0
+                                response = method_to_call(dummy_state_arg)
```
Modified the plugin tester to call the plugin method with an argument.

Myoldmopar commented on testfiles/PythonPluginAirflowNetworkOpeningControlByHumidity.py:
```diff
@@ -9,16 +9,18 @@ def __init__(self):
         self.MyOpenFactor_handle = None
 
-    def on_begin_timestep_before_predictor(self) -> int:
+    def on_begin_timestep_before_predictor(self, state) -> int:
```
The bulk of the changes from here on out are simply converting the plugin python files and the API test files to use the new state argument.

Myoldmopar reviewed (commented):
Walkthrough complete.

mitchute commented on src/EnergyPlus/api/runtime.py:
```diff
@@ -118,7 +121,10 @@ def run_energyplus(self, state: c_void_p, command_line_args: List[Union[str, byt
         return self.api.energyplus(state, len(args_with_program_name), cli_args)
 
-    def issue_warning(self, message: Union[str, bytes]) -> None:
+    # def stop_simulation(self, state: c_void_p) -> None:
```
Keep this around?

Myoldmopar replied:
I should probably keep it and just uncomment it, and wire it up to the underlying C function.  @kbenne 's PR will have some conflicts once this goes in, and while I'm resolving those, I'll uncomment this.  Good eye, though.

mitchute replied:
There's also a couple of others in ```runtime.cc```. FYI

https://github.com/NREL/EnergyPlus/blob/5bf36f34bad6254a899b749d0c76a9d072c25124/src/EnergyPlus/api/runtime.cc#L269-L275

Myoldmopar replied:
Another keen eye. Those are a little different.  The way the User-Defined components work is by taking the name of a program from the IDF and running it explicitly.  This didn't fit well with the callback approach we have for all the other calling points, so these aren't currently enabled.

We should take time after 9.4 to think about how to do component models really well.  For now these will just sit here as commented placeholders.

mitchute reviewed (commented):
No problems from me on this. Unit tests pass locally. I'll attempt to run some other tests by calling E+ as a library.

Myoldmopar committed:
Change state naming conventions in the Python land

Myoldmopar committed:
Added a check to the callback registration to verify function arg count, fixed runtime test

Myoldmopar committed:
Cleanup comments only

Myoldmopar commented:
OK, with these last commits, I cleaned up the state instance class and variable naming so that it is more clear that the object on the API is just an instance of the `StateManager` API category, not an actual instance of an underlying EnergyPlus `state`.  I also added a function to the runtime API class that will validate the number of arguments in the function you are trying to register.  If you try to register a function without any arguments, you will now get an informative TypeError during registration.  I think this will be ready to go in now.

mitchute commented:
Thanks for the cleanup, @Myoldmopar. Clarifying that it's a ```StateManager``` and not an actual instance of ```state``` is great, and the new error message is much better. 

I'm posting my library test file here FFR. 

```python
import sys

# insert path to build so Python can find pyenenergyplus
sys.path.insert(0, "cmake-build-release/Products")

from pyenergyplus.api import EnergyPlusAPI

# setup new API and state instances
api = EnergyPlusAPI()
state = api.state_manager.new_state()


# define method to override calling point
def my_before_timestep_before_predictor(state):
    # get handles
    zone_rh_handle = api.exchange.get_variable_handle(state,
                                                      "System Node Relative Humidity",
                                                      "Zone 1 Node")

    open_factor_handle = api.exchange.get_actuator_handle(state,
                                                          "AirFlow Network Window/Door Opening",
                                                          "Venting Opening Factor",
                                                          "Zn001:Wall001:Win001")

    # get value
    zone_rh = api.exchange.get_variable_value(state, zone_rh_handle)

    # calculate
    if zone_rh < 25.0:
        api.exchange.set_actuator_value(state, open_factor_handle, 0.0)
    elif zone_rh > 60.0:
        api.exchange.set_actuator_value(state, open_factor_handle, 1.0)
    else:
        api.exchange.set_actuator_value(state, open_factor_handle, (zone_rh - 25) / (60 - 25))


# override callback point with my method
api.runtime.callback_begin_system_timestep_before_predictor(state, my_before_timestep_before_predictor)

# execute simulation
idf_path = "testfiles/PythonPluginAirflowNetworkOpeningControlByHumidity.idf"
api.runtime.run_energyplus(state, ["-D", idf_path])
```

To run, remove the ```PythonPlugin:Instance``` object from the ```PythonPluginAirflowNetworkOpeningControlByHumidity.idf``` file and update the paths to the build and test file. 

Anything else you want to do before this merges? If not, I'll get it merged in ASAP.

mjwitte commented:
@Myoldmopar @mitchute Buildling with `LINK_WITH_PYTHON` unchecked fails [here](https://github.com/NREL/EnergyPlus/blob/9123217b2e042a136b6287d5fce1ed5f5e9f86db/src/EnergyPlus/PluginManager.cc#L1492-L1495).

Myoldmopar commented:
Gah!  I forgot CI doesn't always build with that OFF, only on Packages.  I'll fix it directly, please hold.

Myoldmopar commented:
Fixed in develop now, thanks for noting it so quickly @mjwitte !

mjwitte commented:
@Myoldmopar Thanks for the quick fix. Builds nicely here now.
